src/Microsoft.Health.Fhir.CosmosDb/Features/Metrics/CosmosMetricProcessor.cs

```diff
@@ -0,0 +1,119 @@
+        {
+            EnsureArg.IsNotNull(fhirRequestContextAccessor, nameof(fhirRequestContextAccessor));
+            EnsureArg.IsNotNull(mediator, nameof(mediator));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _fhirRequestContextAccessor = fhirRequestContextAccessor;
+            _mediator = mediator;
+            _logger = logger;
+        }
+
+        public void ProcessResponse<T>(T resourceResponseBase)
+            where T : IResourceResponseBase
+        {
+            ProcessResponse(resourceResponseBase.SessionToken, resourceResponseBase.RequestCharge, resourceResponseBase.CollectionSizeUsage, resourceResponseBase.StatusCode);
+        }
+
+        public void ProcessResponse<T>(IFeedResponse<T> feedResponse)
+        {
+            ProcessResponse(feedResponse.SessionToken, feedResponse.RequestCharge, feedResponse.CollectionSizeUsage, statusCode: null);
+        }
+
+        public void ProcessResponse<T>(IStoredProcedureResponse<T> storedProcedureResponse)
+        {
+            ProcessResponse(storedProcedureResponse.SessionToken, storedProcedureResponse.RequestCharge, collectionSizeUsageKilobytes: null, statusCode: storedProcedureResponse.StatusCode);
+        }
+
+        /// <summary>
+        /// Updates the request context with Cosmos DB info and updates response headers with the session token and request change values.
+        /// </summary>
+        /// <param name="sessionToken">THe session token</param>
+        /// <param name="responseRequestCharge">The request charge.</param>
+        /// <param name="collectionSizeUsageKilobytes">The size usage of the Cosmos collection in kilobytes.</param>
+        /// <param name="statusCode">The HTTP status code.</param>
+        public void ProcessResponse(string sessionToken, double responseRequestCharge, long? collectionSizeUsageKilobytes, HttpStatusCode? statusCode)
+        {
+            if (_fhirRequestContextAccessor.FhirRequestContext == null)
+            {
+                return;
+            }
+
+            if (!string.IsNullOrEmpty(sessionToken))
+            {
+                _fhirRequestContextAccessor.FhirRequestContext.ResponseHeaders[CosmosDbHeaders.SessionToken] = sessionToken;
+            }
+
+            AddRequestChargeToFhirRequestContext(responseRequestCharge, collectionSizeUsageKilobytes, statusCode);
+        }
+
+        private void AddRequestChargeToFhirRequestContext(double responseRequestCharge, long? collectionSizeUsage, HttpStatusCode? statusCode)
+        {
+            IFhirRequestContext requestContext = _fhirRequestContextAccessor.FhirRequestContext;
+
+            // If there has already been a request to the database for this request, then we want to append a second charge header.
+            if (requestContext.ResponseHeaders.ContainsKey(CosmosDbHeaders.RequestCharge))
```

Comment:
nit: use TryGetValue here and can use Concat from the existing StringValues below.